app/scripts/modules/core/cluster/filter/clusterFilter.service.ts

```diff
@@ -0,0 +1,475 @@
+    }
+    return shouldInclude;
+  }
+
+  private textFilter(serverGroup: ServerGroup): boolean {
+    const filter: string = this.ClusterFilterModel.sortFilter.filter.toLowerCase();
+    if (!filter) {
+      return true;
+    }
+    if (filter.includes('clusters:')) {
+      const clusterNames: string[] = filter.split('clusters:')[1].replace(/\s/g, '').split(',');
+      return clusterNames.includes(serverGroup.cluster);
+    }
+
+    if (filter.includes('vpc:')) {
+      let vpcName: string = filter.split('vpc:')[1];
+      return serverGroup.vpcName.toLowerCase() === vpcName.toLowerCase();
+    }
+
+    if (filter.includes('tag:')) {
+      let match = false;
+      let [, tag] = filter.split('tag:');
+      let tagKey: string = null;
+      let tagValue: string = null;
+      if (tag.includes('=')) {
+        [tagKey, tagValue] = tag.split('=');
+      }
+      each(serverGroup.tags || {}, (val: string, key: string) => {
+        if (tagKey) {
+          if (tagKey.toLowerCase() === key.toLowerCase() && val.toLowerCase().includes(tagValue.toLowerCase())) {
+            match = true;
+          }
+        } else if (val.toLowerCase().includes(tag.toLowerCase())) {
+          match = true;
+        }
+      });
+      return match;
+    }
+
+    if (filter.includes('detail:')) {
+      let detailName: string = filter.split('detail:')[1];
+      return serverGroup.detail === detailName.toLowerCase();
+    }
+
+    if (filter.includes('cluster:')) {
+      let clusterName: string = filter.split('cluster:')[1];
+      return serverGroup.cluster === clusterName;
+    } else {
+      this.addSearchField(serverGroup);
+      return filter.split(' ').every((testWord: string) => {
+        return serverGroup.searchField.includes(testWord);
+      });
+    }
+  }
+
+  private addSearchField(serverGroup: ServerGroup): void {
+    if (serverGroup.searchField) {
+      return;
+    }
+    let buildInfo = '';
+    if (serverGroup.buildInfo && serverGroup.buildInfo['jenkins']) {
+      buildInfo = [
+        '#' + serverGroup.buildInfo['jenkins']['number'],
+        serverGroup.buildInfo.jenkins.host,
+        serverGroup.buildInfo.jenkins.name].join(' ').toLowerCase();
+    }
+    if (!serverGroup.searchField) {
+      serverGroup.searchField = [
+        serverGroup.region.toLowerCase(),
+        serverGroup.name.toLowerCase(),
+        serverGroup.account.toLowerCase(),
+        buildInfo,
+        (serverGroup.loadBalancers || []).join(' '),
+        (serverGroup.instances || []).map((i: any) => i.id).join(' '),
+      ].join(' ');
+    }
+  }
+
+  private sortGroupsByHeading(groups: IClusterGroup[]): void {
+    this.diffSubgroups(this.ClusterFilterModel.groups, groups);
+    // sort groups in place so Angular doesn't try to update the world
+    this.ClusterFilterModel.groups.sort((a: IClusterGroup, b: IClusterGroup) => {
+      if (a.heading < b.heading) {
```

Comment:
I think this compare function can just be `return a.heading - b.heading;` or drop the `return and `{}`.